File: shared/src/main/java/ru/privatenull/pnauth/message/AuthMessages.java
package ru.privatenull.pnauth.message;

import org.yaml.snakeyaml.Yaml;
import ru.privatenull.pnauth.api.AuthStatus;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public final class AuthMessages {
    private final Map<String, Object> values;
    private final MessageFormat format;
    private final MessageRenderer renderer;

    private AuthMessages(Map<String, Object> values, MessageFormat format) {
        this.values = Map.copyOf(values);
        this.format = format;
        this.renderer = MessageRenderers.forFormat(format);
    }

    public static AuthMessages load(String locale) throws IOException {
        return load(locale, MessageFormat.LEGACY);
    }

    public static AuthMessages load(String locale, MessageFormat format) throws IOException {
        return new AuthMessages(MessageCatalog.defaults(locale), format == null ? MessageFormat.LEGACY : format);
    }

    public static AuthMessages load(Path directory, String locale, MessageFormat format) throws IOException {
        MessageFileGenerator.ensureAll(directory);
        Path file = MessageFileGenerator.ensure(directory, locale);
        Object root = new Yaml().load(Files.readString(file, StandardCharsets.UTF_8));
        Map<String, Object> flattened = new HashMap<>(MessageCatalog.defaults(locale));
        flatten("", root, flattened);
        return new AuthMessages(flattened, format == null ? MessageFormat.LEGACY : format);
    }

    public String text(String key) {
        Object value = values.get(key);
        return renderer.render(value == null ? key : String.valueOf(value));
    }

    public String text(String key, Map<String, String> replacements) {
        Object value = values.get(key);
        return renderer.render(value == null ? key : String.valueOf(value), replacements);
    }

    public List<String> lines(String key) {
        Object value = values.get(key);
        if (value instanceof List<?> list) {
            return list.stream().map(item -> renderer.render(String.valueOf(item))).toList();
        }
        return List.of(text(key));
    }

    public String prompt(AuthStatus status) {
        return text("prompt." + status.name().toLowerCase(Locale.ROOT));
    }

    public MessageFormat format() {
        return format;
    }

    private static InputStream resource(String name) {
        return AuthMessages.class.getResourceAsStream(name);
    }

    private static void flatten(String prefix, Object value, Map<String, Object> output) {
        if (value instanceof Map<?, ?> map) {
            for (Map.Entry<?, ?> entry : map.entrySet()) {
                String key = String.valueOf(entry.getKey());
                flatten(prefix.isEmpty() ? key : prefix + "." + key, entry.getValue(), output);
            }
            return;
        }
        output.put(prefix, value);
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/message/AuthMessages.kt
package ru.privatenull.pnauth.message

import ru.privatenull.pnauth.configuration.SafeYaml
import ru.privatenull.pnauth.api.AuthStatus
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.util.HashMap
import java.util.Locale

class AuthMessages private constructor(
    values: Map<String, Any>,
    val format: MessageFormat
) {
    private val values: Map<String, Any> = java.util.Map.copyOf(values)
    private val renderer: MessageRenderer = MessageRenderers.forFormat(format)

    fun text(key: String): String {
        val value = values[key]
        return renderer.render(value?.toString() ?: key)
    }

    fun text(key: String, replacements: Map<String, String>): String {
        val value = values[key]
        return renderer.render(value?.toString() ?: key, replacements)
    }

    fun lines(key: String): List<String> {
        val value = values[key]
        if (value is List<*>) {
            return value.map { item -> renderer.render(item.toString()) }
        }
        return listOf(text(key))
    }

    fun prompt(status: AuthStatus): String {
        return text("prompt." + status.name.lowercase(Locale.ROOT))
    }

    fun format(): MessageFormat {
        return format
    }

    companion object {
        @JvmStatic
        @Throws(IOException::class)
        @JvmOverloads
        fun load(locale: String, format: MessageFormat? = MessageFormat.LEGACY): AuthMessages {
            return AuthMessages(MessageCatalog.defaults(locale), format ?: MessageFormat.LEGACY)
        }

        @JvmStatic
        @Throws(IOException::class)
        fun load(directory: Path, locale: String, format: MessageFormat?): AuthMessages {
            MessageFileGenerator.ensureAll(directory)
            val file = MessageFileGenerator.ensure(directory, locale)
            val root = SafeYaml.create().load<Any>(Files.readString(file, StandardCharsets.UTF_8))
            val flattened = HashMap<String, Any>(MessageCatalog.defaults(locale))
            flatten("", root, flattened)
            return AuthMessages(flattened, format ?: MessageFormat.LEGACY)
        }

        private fun resource(name: String): InputStream? {
            return AuthMessages::class.java.getResourceAsStream(name)
        }

        private fun flatten(prefix: String, value: Any?, output: MutableMap<String, Any>) {
            if (value is Map<*, *>) {
                for ((k, v) in value) {
                    val key = k.toString()
                    flatten(if (prefix.isEmpty()) key else "$prefix.$key", v, output)
                }
                return
            }
            if (value != null) {
                output[prefix] = value
            }
        }
    }
}
